src/mlpack/tests/augmented_rnns_tasks_test.cpp

```diff
@@ -0,0 +1,281 @@
+    }
+  }
+
+ private:
+  size_t bitLen;
+};
+
+class HardCodedAddModel {
+ public:
+  HardCodedAddModel() {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    return;
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    assert(predictors.n_elem % 3 == 0);
+    predictors = predictors.t();
+    predictors.reshape(3, predictors.n_elem / 3);
+    assert(predictors.n_rows == 3);
+    int num_A = 0, num_B = 0;
+    bool num = false; // true iff we have already seen the separating symbol
+    size_t len = predictors.n_cols;
+    for (size_t i = 0; i < len; ++i) {
+      double digit = arma::as_scalar(arma::find(1 == predictors.col(i), 1));
+      if (digit != 0 && digit != 1)
+      {
+        // We should not see two separators
+        // since we are adding *two* numbers in the task
+        assert(!num);
+        num = true;
+      }
+      else
+      {
+        if (num)
+        {
+          num_B <<= 1;
+          num_B += digit;
+        }
+        else
+        {
+          num_A <<= 1;
+          num_A += digit;
+        }
+      }
+    }
+    int total = num_A + num_B;
+    vector<int> binary_seq;
+    while (total > 0) {
+      binary_seq.push_back(total & 1);
+      total >>= 1;
+    }
+    if (binary_seq.empty()) {
+      assert(num_A + num_B == 0);
+      binary_seq.push_back(0);
+    }
+    size_t tot_len = binary_seq.size();
+    labels = arma::zeros(3, tot_len);
+    for (size_t j = 0; j < tot_len; ++j) {
+      labels.at(binary_seq[tot_len-j-1], j) = 1;
+    }
+    labels.reshape(predictors.n_elem, 1);
+  }
+  void Predict(
+      arma::field<arma::mat>& predictors,
+      arma::field<arma::mat>& labels) {
+    size_t sz = predictors.n_elem;
+    labels = arma::field<arma::mat>(sz);
+    for (size_t i = 0; i < sz; ++i) {
+      Predict(predictors.at(i), labels.at(i));
+    }
+  }
+
+ private:
+  size_t bitLen;
+};
+
+BOOST_AUTO_TEST_SUITE(AugmentedRNNsTasks);
+
+BOOST_AUTO_TEST_CASE(CopyTaskTest)
+{
+  bool ok = true;
+  // Check the setup on vrious lengths...
+  for (size_t maxLen = 2; maxLen <= 16; ++maxLen) {
+    // .. and various numbers of repetitions.
+    for (size_t nRepeats = 1; nRepeats <= 10; ++nRepeats) {
+      CopyTask task(maxLen, nRepeats);
+      arma::field<arma::mat> trainPredictor, trainResponse;
+      task.Generate(trainPredictor, trainResponse, 8);
+      arma::field<arma::mat> testPredictor, testResponse;
+      task.Generate(testPredictor, testResponse, 8);
+      HardCodedCopyModel model;
+      model.Train(trainPredictor, trainResponse);
+      arma::field<arma::mat> predResponse;
+      model.Predict(testPredictor, predResponse);
+      // A single failure is a failure.
+      if (SequencePrecision<arma::mat>(
+            testResponse, predResponse) < 0.99) {
+        ok = false;
+        break;
+      }
+    }
+    if (!ok) break;
+  }
+  BOOST_REQUIRE(ok);
+}
+
+BOOST_AUTO_TEST_CASE(SortTaskTest) {
```

Comment:
If you can add some comments to these tests to make it clear what they are testing, I think that whoever ends up maintaining this code in the future would really appreciate it. :)